### PR TITLE
Fetch Dataset Versions

### PR DESCRIPTION
**Pull request by eddyxu**
Provide `LanceDataset::versions()` and `LanceDataset::latest_version()` methods to fetch available versions.

### CODE REVIEW

File: cpp/src/lance/format/manifest.cc [hunk 1]
```diff
@@ -85,4 +85,6 @@ void Manifest::AppendFragments(const std::vector<std::shared_ptr<DataFragment>>&
   fragments_.insert(std::end(fragments_), std::begin(fragments), std::end(fragments));
 }
 
+arrow::DatasetVersion Manifest::ToDatasetVersion() const { return arrow::DatasetVersion{version_}; }
```
Comment (changhiskhan) [inline cpp/src/lance/format/manifest.cc:88]: nit: not sure what's the right C++ convention but here's we're not turning a Manifest into a DatastVersion, instead we're getting it's version_ as a Datasetversion right? Isn't GetDatasetVersion more appropriate?

File: cpp/src/lance/arrow/dataset.cc [hunk 2]
```diff
@@ -203,28 +214,47 @@ ::arrow::Result<std::shared_ptr<LanceDataset>> LanceDataset::Make(
...
+::arrow::Result<std::vector<DatasetVersion>> LanceDataset::versions() const {
+  std::vector<DatasetVersion> versions;
+  ::arrow::fs::FileSelector selector;
+  selector.base_dir = impl_->versions_dir();
+  selector.allow_not_found = true;
+  selector.recursive = false;
+
+  ARROW_ASSIGN_OR_RAISE(auto file_infos, impl_->fs->GetFileInfo(selector));
+  for (const auto& finfo : file_infos) {
+    ARROW_ASSIGN_OR_RAISE(auto manifest, OpenManifest(impl_->fs, finfo.path()));
```
Comment (changhiskhan) [inline cpp/src/lance/arrow/dataset.cc:231]: Does this need an explicit close or are these automatically cleaned up?

File: cpp/src/lance/arrow/dataset.cc [hunk 3]
```diff
@@ -46,23 +48,22 @@ namespace {
...
-    const std::vector<std::string>& paths, const lance::format::Schema& schema) {
-  std::vector<std::shared_ptr<lance::format::DataFragment>> fragments;
+auto CreateFragments(const std::vector<std::string>& paths, const lance::format::Schema& schema) {
   auto field_ids = schema.GetFieldIds();
-  for (auto path : paths) {
-    fragments.emplace_back(
-        std::make_shared<lance::format::DataFragment>(lance::format::DataFile(path, field_ids)));
-  }
-  return fragments;
+  return paths | views::transform([&field_ids](auto& path) {
```
Comment (changhiskhan) [inline cpp/src/lance/arrow/dataset.cc:62]: just confirming: these don't need any error handling because they're not actually opening the resources right?

**Review by changhiskhan — APPROVED** [review]
will pyarrow.dataset.dataset be able to read this at all? or we need lance.dataset for this to work with versioning?

File: cpp/src/lance/arrow/dataset.cc (hunk 3 above)
Comment (eddyxu) [reply cpp/src/lance/arrow/dataset.cc:62]: Yea, there is no system call in this transform.

File: cpp/src/lance/arrow/dataset.cc (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/arrow/dataset.cc:231]: this opens as `arrow::RandonInputFile` which will be closed once it is out of scope.

File: cpp/src/lance/format/manifest.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/format/manifest.cc:88]: sgtm. will change.

**Comment by eddyxu** [comment]
> will pyarrow.dataset.dataset be able to read this at all? or we need lance.dataset for this to work with versioning?

We need `lance.dataset` until there is a way to reflect `pyarrow.Dataset` from `pyarrow.FileFormat`.

**Pull request merged**